### DISCUSSION TITLE
Complication in collection relations

### DISCUSSION THREAD

**Discussion by brucewayyn** [pocketbase/pocketbase#2212] [Q&A]
Trying to wrap my head around this:

Suppose I have `user`, `cms`, and `post` collections. `cms` should have many `posts`, if a `cms` is deleted, it should delete its related `posts`, but if a `post` is deleted, it should not deleted the related `cms`. A `user` should be able to have many `cms`, and a `cms` should be able to have many `users`.

How can these relations be established, and how should `delete main record on relation delete` for each relation of primary and foreign be set? How does the `primary key` and `foreign key`, and `one-to-many` and `many-to-many` concept work in PocketBase?

Thank you, Gani.

**Comment by ganigeorgiev** [comment]
> how should delete main record on relation delete for each relation of primary and foreign be set?
> How does the primary key and foreign key, and one-to-many and many-to-many concept work in PocketBase?

The questions are too broad and I'm not sure how to answer them.
If you want to assign a relation you have to pass the related record(s) id.
The internal db representation of a record is _*identical_ to the returned object in the record API response (_*the only difference is that array values are stored in the db table as serialized json array_).
There are no db foreign key constraints in the traditional sense in PocketBase since the relations are maintained on app level.
"_Delete main record on relation delete_" is effectively the same as cascade delete. "_main record_" refers to the record in the collection where the `relation` field is defined (_suggestions for better and more clear label are welcome_).

I don't have the full picture of your use case and what `cms` represents, but here is one example structure that I think should satisfy the requirements:

| users (**auth**)      |
| ----------- |
| id      |
| ...   |

| cms (**base**)      |
| ----------- |
| id      |
| users _(multiple `relation` to `users`_     |
| ...   |

| posts (**base**)      |
| ----------- |
| id      |
| cms _(single `relation` to `cms`)_     |
| ...   |

If you enable "_Delete main record on relation delete_" for both relation fields in the above configuration, then:

- on `cms` record delete - all associated `posts` records will be deleted

- on `users` record delete:
  - all `cms` records (and respectively their `posts`), that have only one `cms.users` id matching the deleted user, will be deleted
  - for all `cms` with more than 1 `users` assigned, we will "unlink"/remove its id from the `cms.users` array (_aka. if previously there was `["user1", "user2"]`, after "user1" delete, the value will be `["user2"]`_)

**Reply by brucewayyn** [reply]
@ganigeorgiev thank you for the detailed answer.

> "Delete main record on relation delete" is effectively the same as cascade delete. "main record" refers to the record in the collection where the relation field is defined (suggestions for better and more clear label are welcome).

How does this sound to you:
"Delete record if related record in parent collection is deleted"

<br/>

> - on `users` record delete:
>   - all `cms` records (and respectively their `posts`), that have only one `cms.users` id matching the deleted user, will be deleted

Let me give a scenario: `user1` created `cms1` and added `user2` as author in `cms1` who can also add `posts`. If `user1` is deleted so will `cms1` be deleted because he is the creator of `cms1`, but if `user2` is deleted, `cms1` shouldn't be deleted, he should only be unlinked from `cms1`. If the explanation is clear, how can this be achieved?

And also, what does `listing UI` mean in the below image? I know what the fields achieve, I am not clear on the meaning of `listing UI` here:

<img width="691" alt="Screenshot 2023-04-05 at 5 19 34 PM" src="https://user-images.githubusercontent.com/122556307/230093692-e43d20be-8e78-47a3-8144-187a571bae64.png">

**Reply by ganigeorgiev** [reply]
> If user1 is deleted so will cms1 be deleted because he is the creator of cms1, but if user2 is deleted, cms1 shouldn't be deleted, he should only be unlinked from cms1. 

I'm not sure that I understand what do you mean. If you want a `cms` record to have only one owner then the relation should be single.

> And also, what does listing UI mean in the below image?

Display fields" are only used in the Admin UI (aka. "listing UI" in the above refers to the Admin UI table listing; I'll consider making it more clear in the next release).

**Reply by brucewayyn** [reply]
> I'm not sure that I understand what do you mean. If you want a `cms` record to have only one owner then the relation should be single.

But how can he add another user to the `cms`? In other words, a `cms` should have one owner and many users who can make posts in the `cms`.
